### PR TITLE
Turns Cat Butcher/Surgeon/Whatever into a Basic Mob

### PR DESCRIPTION
**Pull request by san7890**
## About The Pull Request

On the tin, I keep trying to tackle the larger basic mob projects but burn out, so just enjoy these small ones to just keep that list looking lighter and less insurmountable.

No big complex AI either, it choppa the tail off. That's all you really need to be honest.
## Why It's Good For The Game

The cat surgeon will be just that slightly smarter. I suppose they're still a doctor, just simply infatuated with the thought of cat tails...

image

noooooo my tail!!!
## Changelog
:cl:
code: Cat Surgeons are a bit more smarter about attacking threats to the world around them on their endless pursuit for those felinid tails.
/:cl:

### CODE REVIEW

File: code/modules/mob/living/basic/space_fauna/cat_surgeon.dm [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+	faction = list(FACTION_HOSTILE)
+	status_flags = CANPUSH
+	basic_mob_flags = DEL_ON_DEATH
+
+	ai_controller = /datum/ai_controller/basic_controller/cat_butcherer
+
+/mob/living/basic/cat_butcherer/Initialize(mapload)
+	. = ..()
+	apply_dynamic_human_appearance(src, mob_spawn_path = /obj/effect/mob_spawn/corpse/human/cat_butcher, l_hand = /obj/item/circular_saw, bloody_slots = ITEM_SLOT_GLOVES|ITEM_SLOT_OCLOTHING)
+	AddElement(/datum/element/death_drops, list(/obj/effect/mob_spawn/corpse/human/cat_butcher, /obj/item/circular_saw))
```
Comment (MrMelbert) [inline code/modules/mob/living/basic/space_fauna/cat_surgeon.dm:37]: Bespoke elements need static lists passed, etc etc

**Review by MrMelbert — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/basic/space_fauna/cat_surgeon.dm [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+	status_flags = CANPUSH
+	basic_mob_flags = DEL_ON_DEATH
+
+	ai_controller = /datum/ai_controller/basic_controller/cat_butcherer
+
+	/// The stuff we drop on death.
+	var/static/list/drop_on_death = list(
+		/obj/effect/mob_spawn/corpse/human/cat_butcher,
+		/obj/item/circular_saw,
+	)
```
Comment (LemonInTheDark) [inline code/modules/mob/living/basic/space_fauna/cat_surgeon.dm:38]: This makes me depressed

File: code/modules/mob/living/basic/space_fauna/cat_surgeon.dm [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+/mob/living/basic/cat_butcherer/Initialize(mapload)
+	. = ..()
+	apply_dynamic_human_appearance(src, mob_spawn_path = /obj/effect/mob_spawn/corpse/human/cat_butcher, l_hand = /obj/item/circular_saw, bloody_slots = ITEM_SLOT_GLOVES|ITEM_SLOT_OCLOTHING)
+	AddElement(/datum/element/death_drops, drop_on_death)
+	RegisterSignal(src, COMSIG_HOSTILE_POST_ATTACKINGTARGET, PROC_REF(after_attack))
+
+/mob/living/basic/cat_butcherer/Destroy()
+	UnregisterSignal(src, COMSIG_HOSTILE_POST_ATTACKINGTARGET)
```
Comment (LemonInTheDark) [inline code/modules/mob/living/basic/space_fauna/cat_surgeon.dm:47]: Redundant, we clear signals on destroy anyhow

File: code/modules/mob/living/basic/space_fauna/cat_surgeon.dm [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+	return ..()
+
+/mob/living/basic/cat_butcherer/proc/after_attack(mob/living/basic/attacker, atom/target)
+	SIGNAL_HANDLER
+
+	if(!iscarbon(target) && prob(65))
+		return // only a 35% chance of dismemberment
```
Comment (LemonInTheDark) [inline code/modules/mob/living/basic/space_fauna/cat_surgeon.dm:54]: I think your bool logic is messed up. this check passes if the target is a carbon, OR (the target is non carbon AND we pass a 35% chance)
```suggestion
	if(!iscarbon(target) || !prob(35))
		return
```

File: code/modules/mob/living/basic/space_fauna/cat_surgeon.dm [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+	var/mob/living/carbon/human/attacked = target
+	var/obj/item/organ/external/tail/cat/tail = attacked.get_organ_by_type(/obj/item/organ/external/tail/cat)
+	if(!QDELETED(tail))
+		visible_message(
+			span_warning("[src] severs [attacked]'s tail off in one swift swipe!"),
+			span_warning("You sever [attacked]'s tail off."),
+		)
+		tail.Remove(attacked)
+		var/obj/item/organ/external/tail/cat/dropped_tail = new(target.drop_location())
+		dropped_tail.color = attacked.hair_color
```
Comment (LemonInTheDark) [inline code/modules/mob/living/basic/space_fauna/cat_surgeon.dm:65]: Why qdeleted here? when would that ever pass?
Also what the fuck is going on with the tail here? we remove the one they have, make a new one and drop it to the floor.
What happens to the old one?
```suggestion
	if(QDELETED(tail))
		return
	visible_message(
		span_warning("[src] severs [attacked]'s tail off in one swift swipe!"),
		span_warning("You sever [attacked]'s tail off."),
	)
	tail.Remove(attacked)
	var/obj/item/organ/external/tail/cat/dropped_tail = new(target.drop_location())
	dropped_tail.color = attacked.hair_color
```

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/basic/space_fauna/cat_surgeon.dm (hunk 4 above)
Comment (san7890) [reply code/modules/mob/living/basic/space_fauna/cat_surgeon.dm:54]: this is another one of those things from the simple animal implementation but i figure a short-circuit will be faster anyhow

Comment (LemonInTheDark) [reply code/modules/mob/living/basic/space_fauna/cat_surgeon.dm:54]: you short circuited wrong 🗞️ 

x && y && z => !x || !y || !z

File: code/modules/mob/living/basic/space_fauna/cat_surgeon.dm (hunk 5 above)
Comment (san7890) [reply code/modules/mob/living/basic/space_fauna/cat_surgeon.dm]: alright after a bit of research, i think it's just a shitcode thing. `Remove()` on organs operates agnostically, it expects the invoker to do later work with the tail. I have no clue what was going on beforehand with this, but it's fixed now

File: code/modules/mob/living/basic/space_fauna/cat_surgeon.dm (hunk 4 above)
Comment (LemonInTheDark) [reply code/modules/mob/living/basic/space_fauna/cat_surgeon.dm]: explained via discord

File: code/modules/mob/living/basic/space_fauna/cat_surgeon.dm [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+	if(QDELETED(tail))
+		return
+
+	visible_message(
+		span_warning("[src] severs [attacked]'s tail off in one swift swipe!"),
+		span_warning("You sever [attacked]'s tail off."),
+	)
+	tail.Remove(attacked)
+	tail.forceMove(drop_location())
+	dropped_tail.color = attacked.hair_color
```
Comment (LemonInTheDark) [inline code/modules/mob/living/basic/space_fauna/cat_surgeon.dm:63]:
```suggestion
	tail.color = attacked.hair_color
```
Fixes

Comment (san7890) [reply code/modules/mob/living/basic/space_fauna/cat_surgeon.dm]: wtf did vsc not save i hate it here

**Pull request merged**